mixnet/node/src/lib.rs

```diff
@@ -77,205 +76,259 @@ impl MixnetNode {
-        max_retries: usize,
-        retry_delay: Duration,
-        body: Body,
-        to: NymNodeRoutingAddress,
-    ) -> Result<()> {
+    async fn forward(&self, msg: Body, to: NymNodeRoutingAddress) -> Result<()> {
         let addr = SocketAddr::from(to);
-        let arc_socket = pool.get_or_init(&addr).await?;
-
-        if let Err(e) = {
-            let mut socket = arc_socket.lock().await;
-            body.write(&mut *socket).await
-        } {
-            tracing::error!("Failed to forward packet to {addr} with error: {e}. Retrying...");
-            return mixnet_protocol::retry_backoff(
-                addr,
-                max_retries,
-                retry_delay,
-                body,
-                arc_socket,
-            )
+
+        self.message_tx.send(Message::new(addr, msg))?;
+        Ok(())
+    }
+}
+
+struct PacketForwarder {
+    config: MixnetNodeConfig,
+    message_rx: mpsc::UnboundedReceiver<Message>,
+    message_tx: mpsc::UnboundedSender<Message>,
+    connections: HashMap<SocketAddr, TcpStream>,
+}
+
+impl PacketForwarder {
+    pub fn new(
+        message_tx: mpsc::UnboundedSender<Message>,
+        message_rx: mpsc::UnboundedReceiver<Message>,
+        config: MixnetNodeConfig,
+    ) -> Self {
+        Self {
+            message_tx,
+            message_rx,
+            connections: HashMap::with_capacity(config.connection_pool_size),
+            config,
+        }
+    }
+
+    pub async fn run(mut self) {
+        loop {
+            tokio::select! {
+                msg = self.message_rx.recv() => {
+                    if let Some(msg) = msg {
+                        self.handle_msg(msg).await;
+                    } else {
+                        // Channel closed, we should shutdown the message handler thread
+                        return;
+                    }
+                },
+                _ = tokio::signal::ctrl_c() => {
+                    tracing::info!("Shutting down message handler thread...");
+                    return;
+                }
+            }
+        }
+    }
+
+    async fn try_send(&mut self, target: SocketAddr, body: &Body) -> Result<()> {
+        if let std::collections::hash_map::Entry::Vacant(e) = self.connections.entry(target) {
+            match TcpStream::connect(target).await {
+                Ok(tcp) => {
+                    e.insert(tcp);
+                }
+                Err(e) => {
+                    tracing::error!("failed to connect to {}: {e}", target);
+                    return Err(MixnetNodeError::Protocol(e.into()));
+                }
+            }
+        }
+        body.write(self.connections.get_mut(&target).unwrap())
             .await
-            .map_err(Into::into);
+            .map_err(From::from)
+    }
+
+    async fn handle_msg(&mut self, msg: Message) {
+        self.send(msg.target, msg.body, msg.retry_count).await
+    }
+
+    async fn send(&mut self, target: SocketAddr, body: Body, retry_count: usize) {
```

Comment:
Since we have the `Message` struct, what do you think about having only the `send` function which accepts a `Message`? Also, I think we can simplify the signature of the `handle_retry` function below in the same way.

Reply:
I have changed the signature, but do not know why GitHub does not show the changes.